zcash_primitives/src/primitives.rs

```diff
@@ -255,6 +313,12 @@ impl<E: JubjubEngine> Note<E> {
         // Calculate the note contents, as bytes
         let mut note_contents = vec![];
 
+        // Write the asset type
+        self.asset_type
+            .value_commitment_generator::<E>(params)
```

Comment:
See, here we are already calculating the `value_commitment_generator` before writing the note contents.